tensorflow/compiler/mlir/xla/transforms/lhlo_copy_removal.cc

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements a pass to remove redundant LHLO copy operations.
+
+#include "absl/memory/memory.h"
+#include "mlir/Dialect/StandardOps/Ops.h"  // TF:llvm-project
+#include "mlir/IR/Operation.h"             // TF:llvm-project
+#include "mlir/Pass/Pass.h"                // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/passes.h"
+
+namespace mlir {
+namespace xla_hlo {
+namespace {
+
+/// Removes Lhlo.CopyOp that copies from an allocated buffer to the block
+/// argument. All uses of the buffer are replaced with the block argument.
+struct LhloCopyRemoval : mlir::FunctionPass<LhloCopyRemoval> {
+  void runOnFunction() override {
```

Comment:
I think you should ignore functions that are not single block

Actually this could be an Operation Pass, I think this can run on any region that is single block.

Reply:
We took a look at the Operation pass class and realized that it doesn't seem to be possible to manipulate other operations in the current operation pass since we are removing the associated Alloc and Dealloc. We also believe that function pass suits better to our case in order to use live variable analysis later on for solving aliasing problems.

Reply:
I don't understand what you mean by: `it doesn't seem to be possible to manipulate other operations in the current operation pass since we are removing the associated Alloc and Dealloc.`

A Function pass is nothing else than an Operation pass that filters out FuncOp, I don't see what in this pass is depending on the fact that the enclosing operation is a FuncOp in particular?

Reply:
Sorry, I misunderstood your comment. Done.